### DISCUSSION TITLE
Possible bug in simpify

### DISCUSSION THREAD

**Discussion by GitRay** [sympy/sympy#25813] [Q&A]
Before I go running off to the bug tracker, I figured I would ask here for help in case I'm doing something dumb.

Here is a minimal reproducible example, tested on Sympy Live, on v 1.12, and on the master development branch:
```
import sympy
new = '''
Piecewise(
  (1, Eq(z,a)),
  (0, True)
) - 0.5
'''
print(new)
with sympy.evaluate(False):
    new_lhs = sympy.S(new)
    print(new_lhs)
    print(sympy.ccode(new_lhs))
```
The result of this code is:
```
Piecewise(
  (1, Eq(z,a)),
  (0, True)
) - 0.5
Piecewise((1, Eq(z, a)), (0, True)) - 1*0.5
<traceback removed>
TypeError: cannot determine truth value of Relational
```

I believe there are two bugs exposed here. 

1. The first is that when the original expression is run through S(), it is changing the -0.5 to a Mul(-1, 0.5). This would be fine if I wasn't in an evaluate(False) block, because the Mul(-1, 0.5) would (and does) simplify to -0.5. I think the import should be smart enough to avoid creating the unnecessary Mul(-1, 0.5).
2. The second is when the code printer can't handle the Mul being present. 

For anyone who is wondering why the heck I'm using the evaluate(False) block, this is a workaround to prevent sympy from swapping my argument order. Note the difference in the print(new_lhs) line with and without evaluate(False):
```
evaluate(False): Piecewise((1, Eq(z, a)), (0, True)) - 1*0.5
evaluate(True):  Piecewise((1, Eq(a, z)), (0, True)) - 0.5  # a, z is now alphabetical
```

I'm going to start treating this like a bug(s) and trying to fix it unless one of you tell me that I'm being dense :)

**Comment by oscarbenjamin** [comment]
The ccode printer can handle the Mul. What it can't do is run under `evaluate(False)`:
```python
In [36]: new_lhs
Out[36]: 
⎛⎧1  for z = a⎞      
⎜⎨            ⎟ - 0.5
⎝⎩0  otherwise⎠      

In [37]: print(ccode(new_lhs))
((z == a) ? (
   1
)
: (
   0
)) - 0.5
```
Note that many functions will not work under the effect of `evaluate(False)` (gh-25618). 

Ideally you should be able to use `parse_expr` but it does not have the desired effect in this case:
```python
In [38]: parse_expr(new)
Out[38]: 
⎛⎧1  for a = z⎞      
⎜⎨            ⎟ - 0.5
⎝⎩0  otherwise⎠      

In [39]: parse_expr(new, evaluate=False)
Out[39]: 
⎛⎧1  for a = z⎞      
⎜⎨            ⎟ - 0.5
⎝⎩0  otherwise⎠  
```
That is a bug somehow in `parse_expr` or otherwise in `Piecewise`.

Many parsers have the other bug when subtracting numbers:
```python
In [46]: srepr(parse_expr("2 - 3", evaluate=False))
Out[46]: 'Add(Integer(2), Mul(Integer(-1), Integer(3)))'
```
That should be `Add(Integer(2), Integer(-3))` I guess.

**Reply by GitRay** [reply]
Thanks! Maybe I will try to fix this. I have a workaround for this specific case, which is I use rational=True or Rational(1,2) in place of 0.5. 
```
Piecewise((1, Eq(z, a)), (0, True)) - 1*1/2
((z == a) ? (
   1
)
: (
   0
)) - (1.0/2.0)
```
The problem with the proposed deprecation you linked to is that my workaround would no longer work :)

**Reply by oscarbenjamin** [reply]
> The problem with the proposed deprecation you linked to is that my workaround would no longer work

I'm not suggesting the `evaluate(False)` be removed but just that its use should be discouraged and alternatives like `evaluate=False` should be recommended instead. If `evaluate=False` does not work in this case then that is a bug to be fixed somewhere.

**Reply by GitRay** [reply]
The problem is that the printers (in this case ccode) do not accept an "evaluate" keyword - I have no choice but to use "with evaluate(False)". 

Here is an example of the printer reordering the variables when used outside of an evaluate(False) context handler. In this case, the reordering is triggered by the "assign_to" keyword argument:
```
import sympy
new = '''
Piecewise(
  (1, Eq(z,a)),
  (0, True)
)
'''
print(f'new:       {new.strip()}')
with sympy.evaluate(False):
    new_lhs = sympy.S(new, rational=True, evaluate=False)
    print(f'new_lhs:   {new_lhs}')
print(f'ccode: {sympy.ccode(new_lhs)}')
print(f'ccode: {sympy.ccode(new_lhs, assign_to="y")}')
```
Results:
```
new:       Piecewise(
  (1, Eq(z,a)),
  (0, True)
)
new_lhs:   Piecewise((1, Eq(z, a)), (0, True))
ccode: ((z == a) ? (
   1
)
: (
   0
))

ccode: if (a == z) {
   y = 1;
}
else {
   y = 0;
}
```
Am I missing some way to pass "evaluate=False" to the printer?

**Reply by oscarbenjamin** [reply]
> Am I missing some way to pass "evaluate=False" to the printer?

No I don't think so. The change happens here:
https://github.com/sympy/sympy/blob/c226febc80e48826d4314f63a391b31c4d20cfeb/sympy/printing/codeprinter.py#L366-L375
The change is because the Piecewise is reconstructed without passing `evaluate=False`. I don't know whether or not anything depends on the reconstructed Piecewise evaluating here. I imagine not because if the incoming Piecewise was already evaluated then the additional evaluation here would likely not change anything. In that case maybe it is okay to just use `Piecewise(..., evaluate=False)` here.

In general though is there some strong reason that you need to have `z == a` rather than `a == z`?

I know that automatic evaluation can be problematic in this parsing and code printing scenario but reversing the direction of this Eq seems not very problematic.

**Reply by GitRay** [reply]
This gave me a flicker of hope - that I could do my own "Assignment" outside of the call to ccode() but within the "with evaluate(False)" block:
```
import sympy
from sympy.codegen.ast import Assignment
new = '''
Piecewise(
  (1, Eq(z,a)),
  (0, True)
)
'''
print(f'new:       {new.strip()}')
with sympy.evaluate(False):
    new_lhs = sympy.S(new, rational=True, evaluate=False)
    print(f'new_lhs:    {new_lhs}')
    assigned = Assignment(sympy.S('y'), new_lhs)
    print(f'assigned:   {new_lhs}')
print(f'ccode: {sympy.ccode(new_lhs)}')
print(f'ccode: {sympy.ccode(assigned)}')
```
Now this outputs:
```

new:       Piecewise(
  (1, Eq(z,a)),
  (0, True)
)
new_lhs:    Piecewise((1, Eq(z, a)), (0, True))
assigned:   Piecewise((1, Eq(z, a)), (0, True))
ccode: ((z == a) ? (
   1
)
: (
   0
))
ccode: if (a == z) {
   y = 1;
}
else {
   y = 0;
}
```
I'm editing my comment because I previously made an error. It looks like doing the Assignment outside of ccode does in fact respect the evaluate context, and ccode properly prints and "assignment" object - but the reordering still occurs. Thanks for pointing me in the right direction.

I would like the order to stay `z == a` because the generated code is managed by git. Older versions of Sympy conserved the order, so I have a lot of code checked in already with that ordering. My current workflow is to keep track of the Sympy/Python version that each source file was created with, and then when I need to regenerate the equations, I flip back to the appropriate version. Some of the versions are 6 or 7 years old, though, and I'd like to not use unsupported software if I can help it. I could just rip the band-aid off and replace all of the old code, but this burdens the code reviewers unnecessarily - the real changes are only a few lines, not the thousands that they would need to review.

It would probably be better to incorporate Sympy into the build process so that the results of ccode are not tracked by git, but that introduces a pretty big time penalty as of right now. Perhaps my efforts are better directed towards improving the efficiency of my code :)
